### PR TITLE
Optimize EXP

### PR DESCRIPTION
**Pull request by chfast**
In can speed up EXP by 40%. Please verify.

### CODE REVIEW

**Comment by GitCop** [comment]
Thank you for your contribution! Your commits seem to not adhere to the repository coding standards

 * Commit: 0c96205b561960eae4a92b3b316bb7fdaa9c2d5c
  * Commits must be prefixed with the package(s) they modify

Please check the [contribution guidelines](https://github.com/ethereum/go-ethereum#contribution) for more details.

---
This message was auto-generated by https://gitcop.com

File: core/vm/instructions.go [hunk 1]
```diff
@@ -192,8 +192,18 @@ func opSmod(instr instruction, pc *uint64, env Environment, contract *Contract,
 }
 
 func opExp(instr instruction, pc *uint64, env Environment, contract *Contract, memory *Memory, stack *Stack) {
-	x, y := stack.pop(), stack.pop()
-	stack.push(U256(x.Exp(x, y, Pow256)))
+	base, exponent := stack.pop(), stack.pop()
+	result := big.NewInt(1)
+	for exponent.Cmp(common.Big0) != 0 {
+		if exponent.Bit(0) == 1 {
+			result.Mul(result, base)
```
Comment (obscuren) [inline core/vm/instructions.go]: Please write this out as `U256(base.Mul(base, base))`

File: core/vm/instructions.go [hunk 2]
```diff
@@ -192,8 +192,18 @@ func opSmod(instr instruction, pc *uint64, env Environment, contract *Contract,
...
+			U256(result)
+		}
+		base.Mul(base, base)
+		U256(base)
```
Comment (obscuren) [inline core/vm/instructions.go]: Same as above.

**Comment by GitCop** [comment]
Thank you for your contribution! Your commits seem to not adhere to the repository coding standards

 * Commit: c14324e5647ce9d65257c8043c8647322b1fb250
  * Your commit message body contains a line that is longer than 80 characters

Please check the [contribution guidelines](https://github.com/ethereum/go-ethereum#contribution) for more details.

---
This message was auto-generated by https://gitcop.com

**Comment by karalabe** [comment]
Completely perverted modification, though seems to be about 15% faster than this PR. Just putting this out here for the idea.

```go
const (
	// Perverted constants straight out of big.Int
	// https://golang.org/src/math/big/arith.go
	_m    = ^big.Word(0)
	_logS = _m>>8&1 + _m>>16&1 + _m>>32&1
	_S    = 1 << _logS
	_W    = _S << 3 // word size in bits
)

func opExp(instr instruction, pc *uint64, env Environment, contract *Contract, memory *Memory, stack *Stack) {
	base, exponent := stack.pop(), stack.pop()
	result := big.NewInt(1)

	for _, word := range exponent.Bits() {
		for i := 0; i < _W; i++ {
			if word&1 == 1 {
				U256(result.Mul(result, base))
			}
			U256(base.Mul(base, base))
			word >>= 1
		}
	}
	stack.push(result)
}
```

**Comment by chfast** [comment]
@karalabe It is faster for random data, but you will always execute `U256(base.Mul(base, base))` 256 times because the loop is not terminated when the exponent is already 0.

**Comment by chfast** [comment]
This tries to break the loop earlier, but it has the same speed. We would need more tests to be sure.
```
func opExp(instr instruction, pc *uint64, env Environment, contract *Contract, memory *Memory, stack *Stack) {
    base, exponent := stack.pop(), stack.pop()
    result := big.NewInt(1)

    words := exponent.Bits()
    const wordSizeInBits = int(unsafe.Sizeof(words[0]) * 8)
    numWords := len(words)
    for w, word := range words {
        for i := 0; i < wordSizeInBits; i++ {
            if word&1 == 1 {
                U256(result.Mul(result, base))
            } else if word == 0 && w == numWords - 1 {
                break;
            }
            U256(base.Mul(base, base))
            word >>= 1
        }
    }
    stack.push(result)
}
```

**Comment by chfast** [comment]
It looks you took it over, so you can close this one.

**Pull request closed without merging**